File: DFS/23.merge-k-sorted-lists.go
package leetcode

// @leet start
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func mergeKLists(lists []*ListNode) *ListNode {
	switch len(lists) {
	case 0:
		return nil
	case 1:
		return lists[0]
	case 2:
		return merge2(lists[0], lists[1])
	default:
		mid := len(lists) / 2
		l := mergeKLists(lists[:mid])
		r := mergeKLists(lists[mid:])
		return merge2(l, r)
	}
}

func merge2(a, b *ListNode) *ListNode {
	if a == nil {
		return b
	}
	if b == nil {
		return a
	}
	if a.Val < b.Val {
		a.Next = merge2(a.Next, b)
		return a
	} else {
		b.Next = merge2(a, b.Next)
		return b
	}
}

// @leet end

File: DFS/23.merge-k-sorted-lists.py
from typing import List, Optional
from type import ListNode


# @leet start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        match len(lists):
            case 0:
                return None
            case 1:
                return lists[0]
            case _:
                mid = len(lists) // 2
                return self.merge2(
                    self.mergeKLists(lists[mid:]),
                    self.mergeKLists(lists[:mid]),
                )

    def merge2(self, a, b) -> Optional[ListNode]:
        if a is None:
            return b
        if b is None:
            return a
        if a.val < b.val:
            a.next = self.merge2(a.next, b)
            return a
        else:
            b.next = self.merge2(b.next, a)
            return b


# @leet end

ListNode.__lt__ = lambda a, b: a.val < b.val

class Solution:
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        if not lists:
            return None
        mheap = []
        dummy = ListNode()
        cur = dummy
        for n in lists:
            if n:
                heappush(mheap, n)
        while mheap:
            top = heappop(mheap)
            cur.next = top
            cur = cur.next
            if top.next:
                heappush(mheap, top.next)
        return dummy.next


